fix: disabe auto remove if debug is enabled

Haegi opened the pull request:
fixes: https://github.com/orlangure/gnomock/issues/748

### Changes

This PR changes the behaviour of the `WithDebugMode` option. If it is specified, the container will not be auto removed.

 ##### Tests
- Stop container without `WithDebugMode` -> should be gone
- Stop container with `WithDebugMode` -> should be still visible

orlangure commented:
This is awesome!

I see one issue with this change: `gnomock.Stop()` will now not remove the container, if it was created with debug flag, and this isn't the behaviour I would expect in this case. Containers started with debug mode should remain for debugging purposes in case something doesn't work as expected, but if the test completes and calls `Stop` explicitly, the container should be killed completely. Do you agree?

Haegi commented:
> This is awesome!
> 
> I see one issue with this change: `gnomock.Stop()` will now not remove the container, if it was created with debug flag, and this isn't the behaviour I would expect in this case. Containers started with debug mode should remain for debugging purposes in case something doesn't work as expected, but if the test completes and calls `Stop` explicitly, the container should be killed completely. Do you agree?

Good point. I will take a look next week.
I guess just the container removal logic when calling `gnomock.Stop()` is missing, right?

orlangure commented:
>  I guess just the container removal logic when calling gnomock.Stop() is missing, right?

@Haegi correct, because if the `AutoRemove` flag is always set, there is no need to implement container removal; it is enough to just stop it.

Haegi commented:
> > I guess just the container removal logic when calling gnomock.Stop() is missing, right?
> 
> @Haegi correct, because if the `AutoRemove` flag is always set, there is no need to implement container removal; it is enough to just stop it.

I implemented it in https://github.com/Haegi/gnomock/commit/7f3b01680acf223fead437709b956381c09abad3.
**BUT** super rarely I got some message that the deletion is already in progress. But unfortunately,  I didn't find a way to reproduce it.

orlangure commented on docker.go:
```diff
@@ -438,6 +438,18 @@ func (d *docker) stopContainer(ctx context.Context, id string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	err := d.client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{})
```
I think it would be better to use `Force: true` here, just in case.

Haegi replied:
I added it here: https://github.com/orlangure/gnomock/pull/757/commits/a2b478c58f12c6efb0e3c2e9a18e4feb62bfe54a

orlangure commented on gnomock.go:
```diff
@@ -252,7 +252,7 @@ func (g *g) stop(c *Container) error {
 	}
 
-	return nil
+	return cli.removeContainer(context.Background(), id)
```
Please handle the "in progress" error gracefully (return without error), otherwise - fail with error. The error happens most likely due to the `AutoRemove` flag triggering container removal before the code reaches this line.

Haegi replied:
I take care of it in https://github.com/orlangure/gnomock/pull/757/commits/a2b478c58f12c6efb0e3c2e9a18e4feb62bfe54a

orlangure commented on gnomock_test.go:
```diff
@@ -134,6 +137,58 @@ func TestGnomock_cantStart(t *testing.T) {
+		}),
+	})
+	require.NoError(t, err)
+	require.Len(t, containerList, 0)
```
Nice 😼

orlangure reviewed (commented):
@Haegi, thank you for taking care of container removal. There are a couple of small things I think are still missing. Otherwise this is great😻

orlangure commented:
Hi @Haegi,

Thanks for the updates and sorry for taking long to reply.

I like both ideas, but I think investing in the second one is a bit of an overkill right now. The first point is much more important. It would be great if you could take care of it.

Waiting for the final fixes to finally release a new version with your fixes 😼

Haegi commented:
> Hi @Haegi,
> 
> Thanks for the updates and sorry for taking long to reply.
> 
> I like both ideas, but I think investing in the second one is a bit of an overkill right now. The first point is much more important. It would be great if you could take care of it.
> 
> Waiting for the final fixes to finally release a new version with your fixes 😼

No worries :)
I changed the `stopContainer` logic now to not thrown an error when the container is already gone.
I also had to adapt one test to not use the is already gone error but check it with the client.
But still the `TestPreset_duplicateContainerName` is failing and I am not sure why the originalContainer is supposed to fail when trying to stop it.

orlangure commented on docker.go:
```diff
@@ -432,13 +434,25 @@ func (d *docker) stopContainer(ctx context.Context, id string) error {
+	defer d.lock.Unlock()
+
+	err := d.client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{Force: true})
+	if err != nil && !client.IsErrNotFound(err) && isDeletionAlreadyInProgessError(err, id) {
```
I think `isDeletionAlreadyInProgessError` should be negated just like "not found" error, we want to ignore it instead of failing container removal.

Haegi replied:
Thanks for the hint! I had to change the logic of `isDeletionAlreadyInProgessError`. It seems like it was not really checking for the error. Now I was able to get it right and negate it.

orlangure reviewed (commented):
Thank you!

The test you are mentioning fails because of the same (or very similar) reason that the other test you fixed was failing: the original test expected an error because it was a "not found" error when trying to stop a container that doesn't exist anymore (it gets replaced with a new container with the same name). Now that we ignore the "not found" error the test doesn't get an expected error, so we basically change the expectations and should not expect an error anymore.

I also left one last comment, and unless I'm missing something, fixing it together with the test will complete that work, and I'll prepare a new release. This is awesome!

Haegi commented:
> Thank you!
> 
> The test you are mentioning fails because of the same (or very similar) reason that the other test you fixed was failing: the original test expected an error because it was a "not found" error when trying to stop a container that doesn't exist anymore (it gets replaced with a new container with the same name). Now that we ignore the "not found" error the test doesn't get an expected error, so we basically change the expectations and should not expect an error anymore.
> 
> I also left one last comment, and unless I'm missing something, fixing it together with the test will complete that work, and I'll prepare a new release. This is awesome!

Thanks for the explanation. I removed the error expectation and instead check with the docker client if the container is done.

Unfortunately, I noticed some other side effect. gnomockd doesn't return `500` anymore, when trying to stop an already stopped container. Should I try to change it to `404` instead of `500` or is `200` also fine? But `200` might be a bit misleading.

orlangure commented:
I think `200` is fine, the goal is to make sure a container doesn't exist in the end after this endpoint is called, and it really won't exist. Since `cleaner` was added, `Stop` and `/stop` are optional anyway most of the time. I think this is the last fix, and your hard work will be merged and released. Thank you!😻

Haegi commented:
> I think `200` is fine, the goal is to make sure a container doesn't exist in the end after this endpoint is called, and it really won't exist. Since `cleaner` was added, `Stop` and `/stop` are optional anyway most of the time. I think this is the last fix, and your hard work will be merged and released. Thank you!😻

I just changed the test and I also extracted the container listing by id into some helper function called `ListContainerByID`.
It seems like the github action test is a little bit flaky but it is green now.
Looks like we are finally ready to merge this PR :)

orlangure reviewed (approved):
Amazing job, thank you for the contribution!
